dot/sync/worker_pool.go

```diff
@@ -0,0 +1,288 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package sync
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+const (
+	available byte = iota
+	busy
+	punished
+)
+
+const (
+	punishmentBaseTimeout      = 5 * time.Minute
+	maxRequestsAllowed    uint = 60
+)
+
+type syncTask struct {
+	boundTo  *peer.ID
+	request  *network.BlockRequestMessage
+	resultCh chan<- *syncTaskResult
+}
+
+type syncTaskResult struct {
+	who      peer.ID
+	request  *network.BlockRequestMessage
+	response *network.BlockResponseMessage
+	err      error
+}
+
+type peerSyncWorker struct {
+	status         byte
+	timesPunished  int
+	punishmentTime time.Time
+}
+
+type syncWorkerPool struct {
+	wg            sync.WaitGroup
+	mtx           sync.RWMutex
```

Comment:
Do we need to have the mutex as a field since its already in the condvar?

Reply:
yes, the condvar requires a mutex and we still uses the mutex to avoid race conditions among the worker pool